library/core/src/convert/mod.rs

```diff
@@ -478,6 +478,40 @@ pub trait Into<T>: Sized {
 /// - `From<T> for U` implies [`Into`]`<U> for T`
 /// - `From` is reflexive, which means that `From<T> for T` is implemented
 ///
+/// # When to implement `From`
+///
+/// While there's no technical restrictions on which conversions can be done using
+/// a `From` implementation, the general expectation is that the conversions
+/// should typically be restricted as follows:
+///
+/// * The conversion is *lossless*: it cannot fail and it's possible to recover
+///   the original value.  For example, `i32: From<u16>` exists, where the original
```

Comment:
```suggestion
/// * The conversion is *infallible*: if the conversion can fail, use `TryFrom`
///   instead; don't use a `From` impl that panics.
///
/// * The conversion is *lossless*: semantically, it should not lose or discard
///   information. For example, `i32: From<u16>` exists, where the original
```

Reply:
"possible to recover the original value" is not universally true, because it's fine for a `From` impl to convert to an opaque type that doesn't have a means of converting back, as long as it doesn't semantically lose information.

Reply:
Hmm, opaque is a good point.  I'll have to think about how to phrase it to fit with that.

Sadly, even a bunch of standard library `From` impls actually lose information -- `BinaryHeap<T>: From<Vec<T>>` is my usual example -- but I'm also hoping to write this section as SHOULD not MUST, to be a bit stronger in guidance but maybe then allow some wiggle-room on the edges.

Reply:
@scottmcm The phrasing I wrote uses "should", for that exact reason.

Reply:
`BinaryHeap<T>: From<Vec<T>>` is an interesting example. Semantically this says "here are a bunch of elements whose order I don't care about (but I know the fact that they're contiguously allocated is going to be useful to you)". It's not necessary to be able to exactly recover the original order in order for a `From` impl to be justified, because we have said the order isn't semantically significant. In my view this is similar to `String: From<Cow<str>>`. You cannot recover the information about whether the original Cow was borrowed or owned. But that is not part of the semantic state of the original object.

2 other examples we discussed in the library API in relation to _"and it's possible to recover the original value"_ were:

- `ExitCode: From<u8>`. It's lossless in a sense, but you cannot recover that u8.

- `Stdio: From<File>`. Stdio is an opaque thing that only makes sense to the std::process API, for the purpose of setting up the stdin/stdout/stderr of child processes. You cannot recover that File.

I would not want _"it's possible to recover the original value"_ to be part of this definition even if it's hedged with "should" or "most" or some other way; I think it is not the right intuition for articulating this aspect of the role of `From` conversions.

Josh's wording looks acceptable to me.